### DISCUSSION TITLE
SMS for two-factor authentication (or settings toggle)

### DISCUSSION THREAD

**Discussion by sampl** [npm/feedback#753] [Website] [answered]
Hi all,

I've been getting npm email reminders to install 2fa every once in a while. I don't maintain any packages, but I've tried a few times to set it up (just to not get reminded anymore). Every time I do, I realize I need to install a new app or have a security device to continue and I stop.

Maybe one or more of the following would be nice for folks like me:

1. Don't send reminders to npm accounts without packages, or
2. Support SMS two-factor so I don't have to install a new app to set it up, or
3. Allow me to turn off 2fa reminders in settings (looked but didn't see it)

Side note - resetting my password was quite the ordeal on top of that (lots of re-logging in), but that can be a different thread :)

**Comment by ljharb** [comment]
SMS is not a secure mechanism for 2FA due to simjacking attacks. The Google Authenticator app is a pretty common app available on both phone platforms that should work well.

**Comment by sampl** [comment]
Thanks Jordan.

Any thoughts on the other two points?

**Reply by ljharb** [reply]
I don't work for npm ofc, but for the first point, I agree - but it's pretty rare for an account to exist without the *intention* of publishing packages, so i'm not sure how high a priority that would be to "fix".

For the third, 2FA is fast becoming mandatory, so tbh I think you're just going to have to accept that this is now a world where a TON of sites - including banks and other non-tech-user-serving sites - are going to have 2FA, and you'll often need (or at least want) an Authenticator app for that.

**Comment by MylesBorins** [comment, marked as answer]
> Don't send reminders to npm accounts without packages

Unfortunately this is not something we are going to change. We believe that enabling 2FA is the most important action that an account to take to protect both themselves and those relying on their published packages.

> Support SMS two-factor so I don't have to install a new app to set it up

As mentioned by @ljharb this is not something we are going to be doing. [Sim Swap attacks](https://en.wikipedia.org/wiki/SIM_swap_scam)
 are extremely prevalent and it is not considered to be a reliable second factor. Even Authenticator apps are less than ideal and are fishable.

We support security keys, touch-id, face-id, and other second factors supported by WebAuthn. You don't need to install an app to setup 2FA.

> Allow me to turn off 2fa reminders in settings (looked but didn't see it)

See above, this isn't going to be something we do

**Reply by sonnybsj** [reply]
I would also like to see a setting to turn off 2FA reminders.

It seems inconsistent that 2FA is optional but the reminders to turn it on are mandatory. 

I understand SMS 2FA is less secure, but is it better than nothing? Or is it considered less secure than the default email 2fa? 

There may be a tradeoff between security and users working around burdensome tasks.

2fa is becoming mandatory but it's rare to see SMS-2fa not being allowed.

I don't think it's that rare for an account to exist without the intent to publish.

**Reply by ljharb** [reply]
sms is indeed less secure than email due to sim spoofing.

**Comment by sampl** [comment]
> Unfortunately this is not something we are going to change 

Respect to y'all, do what you gotta do.

> You don't need to install an app to setup 2FA.

WebAuthn doesn't seem to work for me quite yet (iphone/mbp, chrome, no xcode--can post pics if you want) but hopefully it's a matter of time.

Tangentially related, I think a few small things like this add up to a kind of general frustration using the site for me (2fa emails, super long pw reset process, thinking I would just sign in with github and asking for lots of permissions). I understand security is paramount as it should be, but I wonder if just some different phrasing could make npm feel friendlier and easier for everyone.

If the team is generally receptive to UX/usability suggestions, I'd be happy make separate threads w/ video for each confusing flow, or whatever format you want. If not, no worries too.

**Reply by MylesBorins** [reply]
I you can open some threads with confusing flows we'd be happy to look into them, we have been investing heavily in account recovery based items right now.

I'd be particularly interested to understand why WebAuthn 2FA isn't working for you. Do you have a macbook with touchid? Which browser are you using?

**Reply by sampl** [reply]
Not sure if this deserves a new thread, but it's a Macbook without touch id (MacBook Pro 13-inch, 2016) and Chrome (Version 104.0.5112.101). I'm traveling atm but I could check on my mac desktop in a month or so.

Here's the prompt from Chrome (explicitly says Android only)
<img width="693" alt="Screen Shot 2022-08-23 at 8 57 20 AM" src="https://user-images.githubusercontent.com/231851/186164528-ddd2eadf-7823-48d9-8a21-91a4e417525b.png">

<img width="628" alt="Screen Shot 2022-08-23 at 8 57 27 AM" src="https://user-images.githubusercontent.com/231851/186164569-575cbaba-62a3-46a0-acb0-fcc63d1d4941.png">

And here's what happens when I scan it on my phone

![IMG_0263](https://user-images.githubusercontent.com/231851/186164451-112a26a1-30cb-4a3f-9cc5-fa39e4118d0f.PNG)

![IMG_0264](https://user-images.githubusercontent.com/231851/186164476-a6d76f58-18a2-48ae-a645-70a8fbc5596d.PNG)

**Comment by sampl** [comment]
Makes sense, thanks for your time Jordan :)
